Adding interface for ST7066U Parallel LCD Driver

HuangDave opened the pull request:
Added the ST7066U.hpp to the firmware/library/L1Drivers directory.
Included in the header file are the ST7066UInterface and ST7066U driver classes.

HuangDave commented:
Not sure why but the Codacy/PR Quality Review is not passing because the member variables lineNum and pos in the CursorPosition_t are unused.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+
```
Remove extra newline

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+ public:
+    struct CursorPosition_t;
+
+    virtual void init()                                               = 0;
```
Prefer not to abbreviate, write out the whole word for the method name.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+#include "L0_LowLevel/LPC40xx.h"
+#include "L1_Drivers/gpio.hpp"
+
+class ST7066UInterface
```
Please follow Google C++ Style guide for naming variables, methods and classes.
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+class ST7066UInterface
+{
+ public:
+    struct CursorPosition_t;
```
This implementation of a structure should not be apart of the interface, and probably shouldn't be apart of the public section.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+
+    virtual void init()                                               = 0;
+    virtual void clearDisplay()                                       = 0;
+    virtual void toggleDisplay(bool on)                               = 0;
```
This function name is misleading. I see this and think that if I run this function it will toggle the display on or off rather than turn it off or on. Please rename.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+    virtual void init()                                               = 0;
+    virtual void clearDisplay()                                       = 0;
+    virtual void toggleDisplay(bool on)                               = 0;
+    virtual void toggleCursor(bool on)                                = 0;
```
Same as above.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+    virtual void clearDisplay()                                       = 0;
+    virtual void toggleDisplay(bool on)                               = 0;
+    virtual void toggleCursor(bool on)                                = 0;
+    virtual void setCursorDirection(bool increment)                   = 0;
```
The method name and parameter do not tell me exactly what happens when I run setCursorDirection. What does this method do?

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+        kLineAddress1 = 0xC0
+    };
+
+    virtual inline void writeNibble(WriteOperationType operation,
```
Make this and the methods below public, and remove the `inline`

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+
+ protected:
+    // Data transfer operation types.
+    enum class WriteOperationType : uint8_t
```
Prefer to use _t when declaring types. But since this is a class (an enumeration class), there shouldn't be a Type word after it.

kammce replied:
Should also make this public. No need to keep this away from developers.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+    enum class WriteOperationType : uint8_t
+    {
+        kCommandWrite = 0,
+        kDataWrite    = 1
```
The word *Write* is a bit redundant here since the name of the class is WriteOperation. From ourside of the class one would write this: `WriteOperation::kData` or `WriteOperation::kCommand`. Since write is in the enum name, the "write" is understood.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+        kDataWrite    = 1
+    };
+
+    enum class Command : uint8_t
```
Make this public as well.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+    virtual inline void resetCursorPosition()                   = 0;
+};
+
+class ST7066U : public ST7066UInterface
```
This commit should not include implementation details such as this. Save for the next commit.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,196 @@
+
+// The file contains the driver class for the ST7066U Parallel LCD Driver.
```
This usage section is a bit confusing. I would say remove it for now, and add it back in the next commit.

kammce commented:
Unused variables should be deleted and codacy is correct to fail on you for that.

HuangDave commented:
Removed CrusorPosition_t from interface.
Renamed init() to Initialize().
Renamed toggleDisplay(bool on) to SetDisplayOn(bool on).
Renamed toggleCursor(bool on) to SetCursorHidden(bool hidden).
Added a CursorDirection enum class and changed SetCursorDirection(bool increment) to SetCursorDirection(CursorDirection direction).
Renamed WriteOperationType to WriteOperation and removed the Write suffixes.
Moved WriteOperation and Command enum classes to public.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,60 @@
+    virtual void SetCursorDirection(CursorDirection direction)             = 0;
+    virtual void SetCursorPosition(CursorPosition_t position)              = 0;
+    virtual void ResetCursorPosition()                                     = 0;
+    virtual void DisplayText(const char * text, CursorPosition_t position) = 0;
```
CursorPosition_t is not defined in this file.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,60 @@
+    virtual void WriteByte(WriteOperation operation, uint8_t byte)         = 0;
+    virtual void WriteCommand(Command command)                             = 0;
+    virtual void WriteCommand(uint8_t command)                             = 0;
+    virtual void WriteData(uint8_t data)                                   = 0;
```
Also, uint8_t is not defined here. You need to include <cstdint>.

kammce commented on firmware/library/L1_Drivers/ST7066U.hpp:
```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include "L0_LowLevel/LPC40xx.h"
+#include "L1_Drivers/gpio.hpp"
```
Remove these includes. They aren't used in this file.
